### PR TITLE
Pull models out of global scope and upgrade to d3fc 3.0.0

### LINKED ISSUE #304: Remove usage of global models
Model objects should not be created in global scope (such as periods, products and user selections), but should be created at the top level - app.js - then passed into the components below, which can then retrieve sub-component models from the model passed into the components (bound to the selection's data).

### LINKED ISSUE #307: Update d3fc dependency to 3.x.x

### PR DESCRIPTION
**Pull request by jleft**
~~Definitely still a WIP, but I thought this was worth showing as (I think) it works and follows the desired pattern (models not in global scope, but declared and mutated in app.js and passed down to components by binding the model to the selection).~~

Any thoughts welcome :-)

Resolves #304. Resolves #307.

### CODE REVIEW

File: src/assets/js/app.js [hunk 1]
```diff
@@ -137,18 +189,18 @@
...
-                        dataInterface(sc.model.selectedPeriod.seconds);
-                    } else if (product.option === sc.model.product.generated) {
+                    updateModelSelectedProduct(product.option);
+                    updateModelSelectedPeriod(product.option.periods[0]);
+                    if (product.option === bitcoin) {
+                        dataInterface(product.option.periods[0].seconds);
+                    } else if (product.option === generated) {
                         dataInterface.generateDailyData();
                     }
                     render();
```
Comment (djmiley) [inline src/assets/js/app.js]: Can we remove this call to `render`?

File: src/assets/js/chart/legend.js [hunk 2]
```diff
@@ -16,11 +15,18 @@
                 ['volume', function(d) { return formatVolume(d.volume); }]
             ]);
 
+
```
Comment (djmiley) [inline src/assets/js/chart/legend.js]: Line

File: src/assets/js/chart/legend.js [hunk 3]
```diff
@@ -16,11 +15,18 @@
...
         function legend(selection) {
-            if (!selection.datum()) {
-                selection.datum(sc.model.latestDataPoint);
-            }
-            selection.call(legendComponent);
+            selection.each(function(model) {
+                var container = d3.select(this);
```
Comment (djmiley) [inline src/assets/js/chart/legend.js]: What is the reason you're not just calling it on the 'selection'? Or why is it not implemented in the other charts?

Comment (djmiley) [inline src/assets/js/chart/legend.js]: Sorry,forget this.

**Comment by djmiley** [comment]
Yes, looking good at the moment.

File: src/assets/js/app.js (hunk 1 above)
Comment (jleft) [inline src/assets/js/app.js]: We could do, as render would be called via a change in dataInterface, but I think explicitly calling render is clearer.

Comment (tlsim) [inline src/assets/js/app.js]: We're calling `renderInternal` in requestAnimationFrame via `fc.util.render`. Doesn't that make any extra `renderInternal` calls redundant?

Comment (jleft) [inline src/assets/js/app.js]: @tlsim - I agree, but I'm not fully sure I understand the question?

Comment (tlsim) [inline src/assets/js/app.js]: No problem, I should have looked at https://github.com/ScottLogic/d3fc/blob/master/src/util/render.js first.

**Comment by tlsim** [comment]
LGTM :+1:

**Comment by jleft** [comment]
Thanks @djmiley and @tlsim for the review :)

**Pull request merged**